rs/embedders/fuzz/src/compile.rs

```diff
@@ -1,44 +1,133 @@
         config.allow_invalid_funcs = true;
-        Ok(MaybeInvalidModule(Module::new(config, u)?))
+        config.memory_offset_choices = MemoryOffsetChoices(40, 20, 40);
+        Ok(MaybeInvalidModule {
+            module: Module::new(config.clone(), u)?,
+            memory64_enabled: config.memory64_enabled,
+        })
     }
 }
 
 #[inline(always)]
-pub fn run_fuzzer(module: MaybeInvalidModule) {
-    let mut config = EmbeddersConfig::default();
-    config.feature_flags.wasm64 = FlagStatus::Enabled;
-    let wasm = module.0.to_bytes();
-    let binary_wasm = BinaryEncodedWasm::new(wasm);
-    let embedder = WasmtimeEmbedder::new(config, no_op_logger());
-
-    let (_, _) = compile(&embedder, &binary_wasm);
+pub fn run_fuzzer(bytes: &[u8]) {
+    let config;
+    let mut u = Unstructured::new(bytes);
+
+    // Arbitrary Wasm module generation probabilities
+    // 33% - Random bytes
+    // 33% - Wasm with arbitrary wasm-smith config + maybe invalid functions
+    // 33% - IC compliant wasm + maybe invalid functions
+
+    // Only used w/ random bytes
+    let memory64_enabled = u.ratio(1, 2).unwrap_or(false);
+
+    let wasm = if u.ratio(1, 3).unwrap_or(false)
+        || bytes.len() < <MaybeInvalidModule as Arbitrary>::size_hint(0).0
+    {
+        config = ic_embedders_config(memory64_enabled);
+        raw_wasm_bytes(bytes)
+    } else {
+        let data = <MaybeInvalidModule as Arbitrary>::arbitrary_take_rest(u);
+
+        match data {
+            Ok(data) => {
+                config = ic_embedders_config(data.memory64_enabled);
+                data.module.to_bytes()
+            }
+            Err(_) => {
+                config = ic_embedders_config(memory64_enabled);
+                raw_wasm_bytes(bytes)
+            }
+        }
+    };
+
+    let rt: Runtime = tokio::runtime::Builder::new_multi_thread()
+        .worker_threads(3)
+        .max_blocking_threads(1)
+        .enable_all()
+        .build()
+        .unwrap_or_else(|err| panic!("Could not create tokio runtime: {}", err));
+
+    let first_execution = rt.spawn({
+        let wasm = wasm.clone();
+        let binary_wasm = BinaryEncodedWasm::new(wasm);
+        let embedder = WasmtimeEmbedder::new(config.clone(), no_op_logger());
+
+        async move { compile(&embedder, &binary_wasm) }
+    });
+
+    let second_execution = rt.spawn({
+        let binary_wasm = BinaryEncodedWasm::new(wasm);
+        let embedder = WasmtimeEmbedder::new(config, no_op_logger());
+
+        async move { compile(&embedder, &binary_wasm) }
+    });
+
+    rt.block_on(async move {
+        // The omitted field is EmbedderCache(Result<InstancePre<StoreData>, HypervisorError>)
+        // 1. InstancePre<StoreData> doesn't implement PartialEq
+        // 2. HypervisorError is the same in compilation_result which is checked for equality
+
+        let (_, compilation_result_1) = first_execution.await.unwrap();
+        let (_, compilation_result_2) = second_execution.await.unwrap();
```

Comment:
Think running it twice will be enough? Also did you double check that this catches the nondeterminism issue we fixed?